plugins/pulp_rpm/plugins/db/models.py

```diff
@@ -188,6 +193,44 @@ def __cmp__(self, other):
             other.complete_version_serialized
         )
 
+    def get_or_calculate_checksum(self, checksumtype):
+        """
+        Tries to get the sha1 or sha256 checksum value from the attribute on self of the
+        corresponding name. If it is not available, calculates the value and saves it on self
+        before returning it.
+
+        SIDE EFFECT: This method will save the model if the requested checksumtype is not populated
+                     on the model in the "checksums" attribute.
+
+        :param checksumtype:   any checksum type supported by the platform
+        :type  checksumtype:   basestring
+
+        :return:    checksum value of the requested type
+        :rtype:     basestring
+
+        :raises PulpCodedException: if the checksumtype is not available
+        """
+        if checksumtype not in server_util.CHECKSUM_FUNCTIONS:
+            raise ValueError(_('Checksum type %(checksumtype)s is not supported'))
+
+        value = self.checksums.get(checksumtype)
+
+        if not value:
+            if self.checksumtype == checksumtype:
+                value = self.checksum
+            elif not self.downloaded:
+                raise PulpCodedException(error_code=error_codes.RPM1007,
+                                         checksumtype=checksumtype)
+            else:
+                _LOGGER.debug(_('calculating checksum of type %(ctype)s for unit %(unit)s') %
+                              {'ctype': checksumtype, 'unit': self})
+                with open(self._storage_path) as f:
```

Comment:
Fair question. If ``self.downloaded == True`` as verified above, then ``self._storage_path`` better contain a valid path! It wouldn't hurt to be more defensive though. I'll think about what a reasonable course of action would be in the case the file can't be opened.